### PR TITLE
WIP: Using persist.History in Replay

### PR DESCRIPTION
**Pull request by ericdill**
# Things needed by this PR
## Save State in the following models:
- [x] ScanIDSearchModel (07abdb9)
- [x] GetLastModel (07abdb9)
  ~~\- [ ] MuxerModel since the muxer model is based entirely on a run header and we are not saving those in the state DB, i'm not sure what other state is important to save here... ohh maybe the way in which each dataset was being up/downsampled.~~ punting this to #115 

~~\- [ ] ScalarCollection~~ handled by the 'saving plotting state' checkbox

~~\- [ ] DisplayHeaderModel~~ saving state doesn't make sense here
- [x] WatchHeadersModel (11d5bd4)
- [x] Saving plotting state (x, y, x_is_time) on a per-run basis
- [x] Figure out how to make this play nice with the concept of saving state from run to run while viewing live data in replay.  Basically what is the precedence order?  If replay finds some state saved in the persist sql db, does it prefer that over the things that are currently being plotted? Or does the current state supersede the persistence sql db?  The compromise is probably to have a check box that flips the persistence order...  6ffc1c6
## Write tests that evaluate state saving in the following models:
- [x] ScanIDSearchModel https://github.com/NSLS-II/dataportal/commit/491e2c300b74d4d34cc6001a7eb63d322d898b23#diff-411c7a2931ab7af1e420a1042a33d726R304
- [x] GetLastModel https://github.com/NSLS-II/dataportal/commit/491e2c300b74d4d34cc6001a7eb63d322d898b23#diff-411c7a2931ab7af1e420a1042a33d726R306

~~\- [ ] MuxerModel~~ punting to #115 

~~\- [ ] ScalarCollection~~ handled by the 'saving plotting state' checkbox

~~\- [ ] DisplayHeaderModel~~ saving state doesn't make sense here
- [x] WatchHeadersModel https://github.com/NSLS-II/dataportal/commit/491e2c300b74d4d34cc6001a7eb63d322d898b23#diff-411c7a2931ab7af1e420a1042a33d726R302
- [x] Saving plotting state (x, y, x_is_time) on a per-run basis https://github.com/NSLS-II/dataportal/commit/491e2c300b74d4d34cc6001a7eb63d322d898b23#diff-411c7a2931ab7af1e420a1042a33d726R300
- [x] put test coverage on switching from disk state to ram state  d7bfd8d

### CODE REVIEW

File: dataportal/replay/core.py [hunk 1]
```diff
@@ -35,3 +35,14 @@ def generate_grid(container, num_cols):
...
                  column_align='h_center'),
             align('width', *widgets)]
+
+
+non_stateful_attrs = ['history']
+
+def save_state(history, history_key, state):
+    for key, val in state.items():
+        if key in non_stateful_attrs:
+            state[key] = None
```
Comment (tacaswell) [inline dataportal/replay/core.py]: I would do this as `state = {k, v for k, v in state.items() if k not in non_stateful_attrs}`

File: dataportal/replay/scalar/model.py [hunk 2]
```diff
@@ -272,12 +317,12 @@ def get_plotting_state(self):
                  'y': [y for y in self.column_models.values() if y.is_plotting]}
         return state
 
-    def set_plot_state(self, changed):
+    def set_plot_state(self, state_dict):
```
Comment (tacaswell) [inline dataportal/replay/scalar/model.py]: can you make a copy before mutating the input?

File: dataportal/replay/core.py (hunk 1 above)
Comment (ericdill) [inline dataportal/replay/core.py]: Agreed

**Comment by ericdill** [comment]
@tacaswell @danielballan The persistent state bit isn't covered under the test framework, but the rest of it is good to go. Any comments thus far?

**Comment by ericdill** [comment]
@tacaswell @danielballan Can you guys review this?  Tests passed locally

**Comment by ericdill** [comment]
You've got to be kidding me.... you just merged a breaking change to mds under me...

**Comment by danielballan** [comment]
I couldn't tell it was breaking because the replay tests were already
broken. Didn't mean that PR to change anything used by replay. Sorry. Let
me know what I should revert.

**Comment by ericdill** [comment]
I'm just going to merge your branch into mine and make the tests pass then push it onto this PR

**Comment by ericdill** [comment]
WOO YEA TRAVIS PASSED

**Comment by ericdill** [comment]
Note that this supersedes #120

**Comment by tacaswell** [comment]
Did it just require the py3k compat fix?

**Comment by ericdill** [comment]
Yes, my replay tests now pass because @danielballan fixed the things that the PR in mds broke and his replay tests now pass because of my fixes to the replay tests

**Comment by ericdill** [comment]
And so the only thing left was the py3k compat

**Comment by tacaswell** [comment]
yay teamwork

**Comment by ericdill** [comment]
yeah, to be honest, i'm rather surprised that it was that simple... no merge conflicts either

File: dataportal/broker/simple_broker.py [hunk 3]
```diff
@@ -279,63 +285,119 @@ def fill_event(event):
...
+                'beamline_id': 'beamline_id',
+                'owner': 'owner',
+                'group': 'group',
+                'project': 'project',
+                'run_start_id': 'id',
+                'run_start_uid': 'uid'}
+        for new_key, old_key in attrs.items():
+            header[new_key] = run_start[old_key]
+        if run_stop is not None:
+            attrs = {'stop_time': 'time',
```
Comment (tacaswell) [inline dataportal/broker/simple_broker.py]: I had an outstanding complaint about this re-use of a variable.

File: dataportal/broker/simple_broker.py [hunk 4]
```diff
@@ -279,63 +285,119 @@ def fill_event(event):
...
+                        color='lightgrey')
+        s.write("run_start_uid='{0}'".format(self.run_start_uid),
+                'lightgrey')
+        if hasattr(self, 'run_stop_uid'):
+            s.write("run_stop_uid='{0}'".format(self.run_stop_uid),
+                    'lightgrey')
+        return s.readout()
+
+
+class Stream(object):
```
Comment (tacaswell) [inline dataportal/broker/simple_broker.py:386]: and a complaint about the naming + location of this code which @danielballan asked we punt on.

File: dataportal/replay/core.py [hunk 5]
```diff
@@ -35,3 +37,48 @@ def generate_grid(container, num_cols):
...
+        True: remove keys from the state dictionary that appear in the
+            `non_stateful_attrs` module-level list
+    """
+    if blacklist:
+        # remove keys that are not helpful for
+        state = {k: v for k, v in state.items() if not (k in non_stateful_attrs
+                 or isinstance(v, tuple(non_stateful_types)))}
+    if sanitize:
+        # remove objects that cannot be serialized
+        state = json.dumps(state, skipkeys=True)
```
Comment (tacaswell) [inline dataportal/replay/core.py]: I think this will lead to double-encoding.  This line needs to be `json.loads(json.dumps(state, skipkeys=True))`

``` python
In [22]: import json

In [23]: tt = {'a': 1, 'b':2}

In [24]: json.dumps(tt)
Out[24]: '{"a": 1, "b": 2}'

In [25]: json.dumps(json.dumps(tt))
Out[25]: '"{\\"a\\": 1, \\"b\\": 2}"'

In [26]: json.loads(json.dumps(json.dumps(tt)))
Out[26]: '{"a": 1, "b": 2}'

In [27]: json.loads(json.loads(json.dumps(json.dumps(tt))))
Out[27]: {'a': 1, 'b': 2}

In [28]: 
```

File: dataportal/replay/replay.py [hunk 6]
```diff
@@ -98,14 +122,25 @@ def main():
...
-    print('args: {}'.format(args))
+    logger.info('args: {}'.format(args))
     if args.live:
         params_dict = define_live_params()
     elif args.small_screen:
         params_dict = define_small_screen_params()
+    loglevel = logging.WARNING
+    if args.verbose:
+        loglevel = logging.INFO
+    elif args.debug:
```
Comment (tacaswell) [inline dataportal/replay/replay.py:149]: :+1:

File: dataportal/replay/scalar/model.py [hunk 7]
```diff
@@ -253,49 +265,46 @@ def __init__(self):
...
-            print('setting {} to self.{}'.format(v, k))
-            setattr(self, k, v)
-        self.get_new_data_and_plot()
+            logger.debug('getting state for dataframe id: {}'.format(self.dataframe_id))
+            state = self.history.get(six.text_type(self.dataframe_id))
+            logger.debug('state retrieved for dataframe id: {}\nstate: {}'.format(self.dataframe_id, state))
+        except IndexError:
+            # there are no entries in the db for 'state'
+            logger.debug('no state found for dataframe id: {}'.format(self.dataframe_id))
+            state = []
```
Comment (tacaswell) [inline dataportal/replay/scalar/model.py]: should this be an empty dictionary?

File: dataportal/replay/core.py [hunk 8]
```diff
@@ -35,3 +37,48 @@ def generate_grid(container, num_cols):
...
+    ----------
+    history : dataportal.replay.persist.History
+        An instance of the History object which wraps a sql db
+    history_key : str
+        The primary key used to set and retrieve state
+    state : dict
+        Dictionary of state that should be saved to the db
+    sanitize : bool, optional
+        defaults to False
+        True: verify that all values in the state dictionary can be serialized
```
Comment (tacaswell) [inline dataportal/replay/core.py]: This is more of 'strips values that can not be serialized'

File: dataportal/replay/muxer/model.py [hunk 9]
```diff
@@ -299,11 +287,11 @@ def _norm_column_changed(self, changed):
     @observe('data_muxer')
     def _new_muxer(self, changed):
         # data_muxer object has been changed. Remake the columns
-        print('new data muxer received')
+        logger.debug('new data muxer received')
         self._verify_column_info()
 
     def perform_binning(self):
-        print('\n\n\nbinning column: {}'.format(self.binning_column))
+        logger.debug('\n\n\nbinning column: {}'.format(self.binning_column))
```
Comment (tacaswell) [inline dataportal/replay/muxer/model.py]: can drop the `'\n`' in the logging

File: dataportal/replay/replay.py [hunk 10]
```diff
@@ -1,20 +1,32 @@
...
 import argparse
-
+import os
+import logging
 with enaml.imports():
     from dataportal.replay.replay_view import MainView
 
+logger = logging.getLogger(__name__)
+
+REPLAY_CONF_DIR = os.path.join(os.path.expanduser('~'), '.config', 'replay')
```
Comment (tacaswell) [inline dataportal/replay/replay.py]: http://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html

This should be `~/.local/share/replay`

Comment (tacaswell) [inline dataportal/replay/replay.py]: look at the 

> Environment variables
> 
> $XDG_DATA_HOME defines the base directory relative to which user specific data files should be stored. If $XDG_DATA_HOME is either not set or empty, a default equal to $HOME/.local/share should be used.

File: dataportal/replay/replay.py [hunk 11]
```diff
@@ -1,20 +1,32 @@
...
+try:
+    os.mkdir(REPLAY_CONF_DIR)
```
Comment (tacaswell) [inline dataportal/replay/replay.py]: In `filestore` there is a python 2/3 safe version of 'ensure_path'.  I think it is in the numpy writer section.

File: dataportal/replay/replay.py (hunk 10 above)
Comment (tacaswell) [inline dataportal/replay/replay.py]: I just did the same thing is FS.

File: dataportal/replay/replay.py [hunk 12]
```diff
@@ -98,14 +122,25 @@ def main():
...
+
+    replay.logger.setLevel(loglevel)
+    replay.handler.setLevel(loglevel)
 
     # create and show the GUI
-    create(params_dict).show()
+    ui = create(params_dict)
+    ui.show()
+
+    ui.muxer_model.header = db[-1]
```
Comment (tacaswell) [inline dataportal/replay/replay.py]: to be paranoid, wrap this is a try block for when this turns on with an empty mds?

File: dataportal/replay/scalar/model.py [hunk 13]
```diff
@@ -51,25 +52,25 @@ def __init__(self, ax, *args, **kwargs):
     @observe('title')
     def title_changed(self, changed):
         self._ax.set_title(self.title)
-        print('{}: {}'.format(changed['name'], changed['value']))
+        logger.debug('{}: {}'.format(changed['name'], changed['value']))
```
Comment (tacaswell) [inline dataportal/replay/scalar/model.py]: If you use old-style formatting with the loggers like

``` python
logger.debug('%s %s', 'a', 'b')
```

The string formatting will only happen if the log message will really be used, doing it this way the formatting will happen everytime to matter what.

File: dataportal/replay/muxer/model.py [hunk 14]
```diff
@@ -189,40 +195,23 @@ def __init__(self):
...
-            ploty = None
-        else:
-            ploty = [y for y in ploty if y in self.column_models.keys()]
-            plot_state['y'] = ploty
-
-        print('MuxerModel.init_state: {}'.format(plot_state))
-        self.plot_state = plot_state
+        # then change the id
+        # (The scalar_model depends on this order being dataframe then id)
+        self.header_id = self.header.run_start_id
```
Comment (tacaswell) [inline dataportal/replay/muxer/model.py:214]: Might want to do this as `self.dataframe, self.header_id = self.data_muxer._dataframe, self.header.run_start_id` to make sure the order stays fixed.

only 90% that this will work as I expect

Comment (ericdill) [inline dataportal/replay/muxer/model.py:214]: I'm not sure I see any advantage of collapsing it down to a single line

Comment (tacaswell) [inline dataportal/replay/muxer/model.py:214]: Less likely that it well accidentally get re ordered, just mildly paranoid
defensive programming.

File: dataportal/replay/replay.py (hunk 12 above)
Comment (ericdill) [inline dataportal/replay/replay.py]: ~~Works fine when mds is empty~~ nevermind

**Comment by ericdill** [comment]
@tacaswell updated the code based on all but one of your comments

File: dataportal/replay/muxer/model.py (hunk 14 above)
Comment (ericdill) [inline dataportal/replay/muxer/model.py:214]: I mean the in-line comments are pretty clear about the order being important

Comment (ericdill) [inline dataportal/replay/muxer/model.py:214]: Ah, it is also very annoying to make sure that the message passing is working correctly. I really just dont want to spend 30-60 minutes futzing around to make sure that doing it your way didnt break anything
